File: hushh-webapp/android/app/src/main/java/com/hussh/app/plugins/HushhAccount/HushhAccountPlugin.kt
package com.hussh.app.plugins.HushhAccount

import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.hussh.app.plugins.shared.BackendUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MediaType.Companion.toMediaType
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * HushhAccountPlugin - Account Management
 * Port of iOS HushhAccountPlugin.swift
 *
 * Handles account-level operations like deletion.
 */
@CapacitorPlugin(name = "HushhAccount")
class HushhAccountPlugin : Plugin() {

    private val TAG = "HushhAccount"
    private val maxErrorPayloadBytes = 16_384
    private val maxErrorPayloadDepth = 6
    private val maxErrorPayloadNodes = 64
    private val maxErrorPayloadEntries = 32
    private val maxErrorCodeLength = 128
    private val maxErrorMessageLength = 512

    // Longer timeout for deletion operations
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(45, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(45, TimeUnit.SECONDS)
        .build()

    private fun getBackendUrl(call: PluginCall): String {
        return BackendUrl.resolve(bridge, call, "HushhAccount")
    }

    private fun normalizedMachineCode(value: Any?): String? {
        if (value !is String) return null
        val code = value.trim()
        if (code.isEmpty() || code.length > maxErrorCodeLength) return null
        return code.takeIf {
            it.all { character ->
                character == '_' || character in 'A'..'Z' || character in '0'..'9'
            }
        }
    }

    private fun machineCode(
        value: Any?,
        depth: Int,
        remainingNodes: IntArray,
    ): String? {
        if (depth > maxErrorPayloadDepth || remainingNodes[0] <= 0) return null
        remainingNodes[0] -= 1

        if (value is JSONObject) {
            for (key in listOf("code", "error_code")) {
                normalizedMachineCode(value.opt(key))?.let { return it }
            }
            val keys = value.keys()
            var entries = 0
            while (keys.hasNext() && entries < maxErrorPayloadEntries) {
                val code = machineCode(value.opt(keys.next()), depth + 1, remainingNodes)
                if (code != null) return code
                entries += 1
            }
        } else if (value is org.json.JSONArray) {
            val entries = minOf(value.length(), maxErrorPayloadEntries)
            for (index in 0 until entries) {
                val code = machineCode(value.opt(index), depth + 1, remainingNodes)
                if (code != null) return code
            }
        }
        return null
    }

    private fun errorMessage(payload: JSONObject?, fallback: String): String {
        val detail = payload?.opt("detail")
        val detailObject = detail as? JSONObject
        val candidates = listOf(
            detailObject?.opt("message"),
            detailObject?.opt("error"),
            payload?.opt("message"),
            payload?.opt("error"),
            detail,
        )
        for (candidate in candidates) {
            if (candidate !is String) continue
            val message = candidate.trim()
            if (message.isNotEmpty()) return message.take(maxErrorMessageLength)
        }
        return fallback
    }

    private fun parseBoundedErrorPayload(responseBody: String?): JSONObject? {
        if (responseBody.isNullOrBlank()) return null
        if (responseBody.toByteArray(Charsets.UTF_8).size > maxErrorPayloadBytes) return null
        return try {
            JSONObject(responseBody)
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Delete the user's account and user-owned data. Required retained evidence
     * follows the backend's approved retention/redaction policy.
     * Requires VAULT_OWNER token (Unlock to Delete).
     * This action is irreversible.
     */
    @PluginMethod
    fun deleteAccount(call: PluginCall) {
        val authToken = call.getString("authToken")
        if (authToken == null) {
            call.reject("Missing required parameter: authToken")
            return
        }
        val target = call.getString("target") ?: "both"

        val backendUrl = getBackendUrl(call)
        val url = "$backendUrl/api/account/delete"
        val requestBody = JSONObject().apply {
            put("target", target)
        }.toString().toRequestBody("application/json; charset=utf-8".toMediaType())

        Log.w(TAG, "🚨 [HushhAccountPlugin] Requesting account deletion for target: $target")

        Thread {
            try {
                val request = Request.Builder()
                    .url(url)
                    .delete(requestBody)
                    .addHeader("Content-Type", "application/json")
                    .addHeader("Authorization", "Bearer $authToken")
                    .build()

                val response = httpClient.newCall(request).execute()
                val responseBody = response.body?.string()

                activity.runOnUiThread {
                    if (response.isSuccessful) {
                        Log.i(TAG, "✅ [HushhAccountPlugin] Account deleted successfully")
                        val payload = JSObject()
                        var parsed = false
                        try {
                            if (!responseBody.isNullOrBlank()) {
                                val json = JSONObject(responseBody)
                                val keys = json.keys()
                                while (keys.hasNext()) {
                                    val key = keys.next()
                                    payload.put(key, json.get(key))
                                }
                                parsed = true
                            }
                        } catch (_: Exception) {
                            parsed = false
                        }
                        if (!parsed) {
                            payload.put("success", true)
                        }
                        call.resolve(payload)
                    } else {
                        val payload = parseBoundedErrorPayload(responseBody)
                        val remainingNodes = intArrayOf(maxErrorPayloadNodes)
                        val code = machineCode(payload, 0, remainingNodes)
                        val errorMessage = errorMessage(
                            payload,
                            "Server returned ${response.code}",
                        )
                        val rejectionData = JSObject().put("status", response.code)
                        if (payload != null) {
                            rejectionData.put("payload", payload)
                        }
                        Log.e(TAG, "❌ [HushhAccountPlugin] Deletion failed: $errorMessage")
                        call.reject(errorMessage, code, rejectionData)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ [HushhAccountPlugin] Network error: ${e.message}")
                activity.runOnUiThread {
                    call.reject("Network error: ${e.message}")
                }
            }
        }.start()
    }
}

File: hushh-webapp/ios/App/App/Plugins/HushhAccountPlugin.swift
// hushh-webapp/ios/App/App/Plugins/HushhAccountPlugin.swift
import Capacitor

/**
 * HushhAccountPlugin
 * Handles account-level operations like deletion.
 */
@objc(HushhAccountPlugin)
public class HushhAccountPlugin: CAPPlugin, CAPBridgedPlugin {
    
    public let identifier = "HushhAccountPlugin"
    public let jsName = "HushhAccount"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "deleteAccount", returnType: CAPPluginReturnPromise)
    ]
    
    private lazy var urlSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 45 // Longer timeout for deletion
        return URLSession(configuration: config)
    }()

    private let maxErrorPayloadBytes = 16_384
    private let maxErrorPayloadDepth = 6
    private let maxErrorPayloadNodes = 64
    private let maxErrorPayloadEntries = 32
    private let maxErrorCodeLength = 128
    private let maxErrorMessageLength = 512

    private func normalizedMachineCode(_ value: Any?) -> String? {
        guard let raw = value as? String else { return nil }
        let code = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !code.isEmpty, code.count <= maxErrorCodeLength else { return nil }
        let allowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_")
        return code.unicodeScalars.allSatisfy(allowed.contains) ? code : nil
    }

    private func machineCode(
        in value: Any?,
        depth: Int = 0,
        remainingNodes: inout Int
    ) -> String? {
        guard depth <= maxErrorPayloadDepth, remainingNodes > 0 else {
            return nil
        }
        remainingNodes -= 1

        if let dictionary = value as? [String: Any] {
            for key in ["code", "error_code"] {
                if let code = normalizedMachineCode(dictionary[key]) {
                    return code
                }
            }
            for (_, child) in dictionary.prefix(maxErrorPayloadEntries) {
                if let code = machineCode(
                    in: child,
                    depth: depth + 1,
                    remainingNodes: &remainingNodes
                ) {
                    return code
                }
            }
        } else if let array = value as? [Any] {
            for child in array.prefix(maxErrorPayloadEntries) {
                if let code = machineCode(
                    in: child,
                    depth: depth + 1,
                    remainingNodes: &remainingNodes
                ) {
                    return code
                }
            }
        }
        return nil
    }

    private func errorMessage(from payload: [String: Any]?, fallback: String) -> String {
        let detail = payload?["detail"]
        let detailObject = detail as? [String: Any]
        let candidates: [Any?] = [
            detailObject?["message"],
            detailObject?["error"],
            payload?["message"],
            payload?["error"],
            detail,
        ]
        for candidate in candidates {
            guard let raw = candidate as? String else { continue }
            let message = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !message.isEmpty {
                return String(message.prefix(maxErrorMessageLength))
            }
        }
        return fallback
    }
    
    @objc func deleteAccount(_ call: CAPPluginCall) {
        // Get auth token passed from JS layer
        guard let authToken = call.getString("authToken") else {
             call.reject("Missing required parameter: authToken")
             return
        }
        let target = call.getString("target") ?? "both"
        
        let backendUrl = HushhProxyClient.resolveBackendUrl(
            call: call,
            plugin: self,
            jsName: jsName
        )
        
        guard let url = URL(string: "\(backendUrl)/api/account/delete") else {
            call.reject("Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "target": target
        ])
        
        print("🚨 [HushhAccountPlugin] Requesting account deletion for target: \(target)")
        
        let task = urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                call.reject("Network error: \(error.localizedDescription)")
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                if (200...299).contains(httpResponse.statusCode) {
                    if let data = data,
                       let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                        call.resolve(json)
                    } else {
                        call.resolve(["success": true])
                    }
                } else {
                    let boundedData = data.flatMap {
                        $0.count <= self.maxErrorPayloadBytes ? $0 : nil
                    }
                    let payload = boundedData.flatMap {
                        try? JSONSerialization.jsonObject(with: $0) as? [String: Any]
                    }
                    var remainingNodes = self.maxErrorPayloadNodes
                    let code = self.machineCode(
                        in: payload,
                        remainingNodes: &remainingNodes
                    )
                    let message = self.errorMessage(
                        from: payload,
                        fallback: "Server returned \(httpResponse.statusCode)"
                    )
                    var rejectionData: [String: Any] = [
                        "status": httpResponse.statusCode
                    ]
                    if let payload {
                        rejectionData["payload"] = payload
                    }
                    call.reject(message, code, nil, rejectionData)
                }
            } else {
                call.reject("Invalid response")
            }
        }
        task.resume()
    }
}
